### ISSUE TITLE
Private Container reference being switched - crazy!

### ISSUE THREAD

**Issue by shaulbehr** [structuremap/structuremap#517]
Hi,

A while ago, I was experiencing some really wacked out cross-thread conflicts in StructureMap.  I posted on [StackOverflow](http://stackoverflow.com/q/38775306/7850), and eventually [solved](http://stackoverflow.com/a/38814723/7850) my own problem with a workaround.

But this is now coming back to bite us in another part of our code.
We're getting exceptions indicating that the Container has already been disposed, which is basically impossible.  And it only happens under high load conditions.  This makes me suspect that something similar to my previous SO question is now happening in our ExceptionHandlerMiddleware class, where we're getting an exception on line 49, trying to get an instance from the container and finding that `_container` has already been disposed.

Any ideas how this might be possible, and why it might help to switch the reference to local rather than using the private variable to hold the container?

We are using StructureMap 4.4.1

**Comment by shaulbehr** [comment]
[ExceptionHandlerMiddleware.zip](https://github.com/structuremap/structuremap/files/555412/ExceptionHandlerMiddleware.zip)

**Comment by jeremydmiller** [comment]
@shaulbehr If you're up on SM 4.4.1, you can use the `IContainer.DisposalLock` property right after app startup to latch it. Doesn't solve the underlying problem of what in the world is disposing it, but it will stop the problem from taking down your production app.

Just make sure you remove the lock *before* you try to shut down the app so Dispose() can work at the end;)

**Comment by shaulbehr** [comment]
@jeremydmiller I thought through what you're saying, and I don't think the DisposalLock will help.  To be clear about what is happening, here's the logging code and output, copied from the SO answer:
```
_logger.Debug($"Line 1 Context={context.GetHashCode}, Container={Container.GetHashCode()}");
var db = MyDbContext.ForShard(shardKey.Value);  // no need for "using", since DI will automatically dispose
_logger.Debug($"Line 2 Context={context.GetHashCode}, Container={Container.GetHashCode()}");
Container.Configure(cfg => cfg.For<MyDbContext>().Use(db));
await Next.Invoke(context);
```
And here's what came out:
```
Line 1 Context=56852305, Container=48376271
Line 1 Context=88275661, Container=85736099
Line 2 Context=56852305, Container=85736099
Line 2 Context=88275661, Container=85736099
```
See, the reference of the `Container` variable is being switched to a container from another thread!  I don't even know how that's *possible*, unless you've got some kind of singleton architecture happening behind the scenes.  

Being that this is the observed behavior, it won't help to put a disposal lock on Container `48376271`, because by the time the error is thrown, we're looking at Container `85736099`.  And if we also have a disposal lock on Container `85736099`, then we'll never be able to dispose it, because the point in code where it was disposed is where you would *legitimately* expect it to be disposed.

In short, the problem is not that the container is being disposed.  The problem is that the private reference to the container is being *switched* to refer to another container, from another thread, that was *legitimately* disposed.

**Comment by jeremydmiller** [comment]
@shaulbehr I don't know what to tell you at this point. There's nothing in StructureMap itself that would be doing that. What's the application framework you're using? Are you using any kind of custom SM adapter?

**Comment by jeremydmiller** [comment]
@shaulbehr Is this still an issue? Can I close this one yet?

**Comment by shaulbehr** [comment]
It has not been resolved, but I have worked around it.  I guess that means you can close it, but that doesn't imply that the root problem has been solved... ;)
